File: app/src/main/java/com/jianbao/jamboblesdk/OxiMeterActivity.java
package com.jianbao.jamboblesdk;

import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.creative.FingerOximeter.FingerOximeter;
import com.creative.FingerOximeter.IFingerOximeterCallBack;
import com.creative.base.BaseDate;
import com.jianbao.fastble.JamBoHelper;
import com.jianbao.fastble.data.BleDevice;
import com.jianbao.jamboble.BleState;
import com.jianbao.jamboble.callbacks.BleDataCallback;
import com.jianbao.jamboble.callbacks.IBleStatusCallback;
import com.jianbao.jamboble.data.BTData;
import com.jianbao.jamboble.data.OximeterData;
import com.jianbao.jamboble.data.SpO2Data;
import com.jianbao.jamboble.device.BTDevice;
import com.jianbao.jamboble.device.oximeter.OxiMeterHelper;
import com.jianbao.jamboble.device.oximeter.OximeterDevice;
import com.jianbao.jamboble.device.oximeter.OximeterReader;
import com.jianbao.jamboble.device.oximeter.OximeterWriter;
import com.jianbao.jamboble.draw.DrawThreadNW;

import org.jetbrains.annotations.NotNull;

import java.lang.ref.WeakReference;
import java.util.List;

import static com.jianbao.jamboble.device.oximeter.FingerOximeterCallback.MSG_DATA_SPO2_PARA;
import static com.jianbao.jamboble.device.oximeter.FingerOximeterCallback.MSG_DATA_SPO2_WAVE;
import static com.jianbao.jamboble.device.oximeter.FingerOximeterCallback.MSG_PROBE_OFF;

public class OxiMeterActivity extends AppCompatActivity {
    private DrawThreadNW mDtBloodOx;
    private LinearLayout mLayoutMeasuringGuide;
    private TextView mTvValueRealtime;
    private Button mBtnOpenBle;
    private TextView mTvStatus;

    private FingerOximeter fingerOximeter;
    private OximeterWriter oximeterWriter;
    private OximeterReader oximeterReader;

    private Handler handler;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_oxi_meter);

        setTitle("血氧测量");

        mDtBloodOx = findViewById(R.id.dt_blood_ox);
        mLayoutMeasuringGuide = findViewById(R.id.layout_measuring_guide);
        mTvValueRealtime = findViewById(R.id.tv_value_realtime);
        mBtnOpenBle = findViewById(R.id.btn_open_ble);
        mTvStatus = findViewById(R.id.tv_status);

        JamBoHelper.getInstance().setBleDataCallBack(new BleDataCallback() {
            @Override
            public void onBTStateChanged(@NotNull BleState state) {
                switch (state) {
                    //未开启蓝牙
                    case NOT_FOUND:
                        mTvStatus.setText("请打开蓝牙");
                        break;
                    //正在扫描
                    case SCAN_START:
                        mBtnOpenBle.setText("停止扫描");
                        mTvStatus.setText("开始扫描...");
                        break;
                        //连接成功
                    case CONNECTED:
                        mTvStatus.setText("连接设备成功");
                        mLayoutMeasuringGuide.setVisibility(View.INVISIBLE);
                        break;
                    //长时间未搜索到设备
                    case TIMEOUT:
                        mBtnOpenBle.setText("开始扫描");
                        mTvStatus.setText("超时");
                        break;
                    case CONNECTING:
                        mBtnOpenBle.setText("开始扫描");
                        mTvStatus.setText("连接中");
                        break;
                    case DISCONNECT:
                        mDtBloodOx.cleanWaveData();
                        pauseRecord();
                        if (oximeterReader != null) {
                            oximeterReader.close();
                        }
                        if (oximeterWriter != null) {
                            oximeterWriter.close();
                        }
                        JamBoHelper.getInstance().onBTStateChanged(BleState.SCAN_START);
                        mTvStatus.setText("断开连接");
                        break;
                    case CONNECT_FAILED:
                        mTvStatus.setText("连接失败");
                        break;
                    case SCAN_STOP:
                        mBtnOpenBle.setText("开始扫描");
                        break;
                    default:
                        break;
                }
            }

            @Override
            public void onBTDataReceived(@Nullable BTData data) {
                if (data instanceof OximeterData) {
                    BTDevice device = JamBoHelper.getInstance().getBtDevice();
                    if (device instanceof OximeterDevice) {
                        ((OximeterDevice) device).getOximeterHelper().addBuffer(((OximeterData) data).getData());
                    }
                }
            }

            @Override
            public void onLocalBTEnabled(boolean enabled) {

            }
        });

        JamBoHelper.getInstance().setBleStatusCallback(new IBleStatusCallback() {
            @Override
            public void onBTDeviceFound(List<BleDevice> list) {

            }

            @Override
            public void onNotification() {
                BTDevice device = JamBoHelper.getInstance().getBtDevice();
                if (device instanceof OximeterDevice) {
                    OxiMeterHelper helper = ((OximeterDevice) device).getOximeterHelper();
                    oximeterReader = new OximeterReader(helper);
                    oximeterWriter = new OximeterWriter(helper);
                    fingerOximeter = new FingerOximeter(oximeterReader, oximeterWriter,
                            new IFingerOximeterCallBack(){

                                public void OnGetSpO2Param(int nSpO2, int nPR, float fPI, boolean nStatus, int nMode, float nPower, int powerLevel) {
                                    SpO2Data mSpO2Data = new SpO2Data();
                                    mSpO2Data.setSpO2(nSpO2);
                                    mSpO2Data.setPR(nPR);
                                    mSpO2Data.setPI(fPI);
                                    mSpO2Data.setStatus(nStatus);
                                    mSpO2Data.setMode(nMode);
                                    mSpO2Data.setPower(nPower);
                                }

                                @Override
                                public void OnGetSpO2Wave(List<BaseDate.Wave> list) {

                                }

                                @Override
                                public void OnGetDeviceVer(String s, String s1, String s2) {

                                }

                                @Override
                                public void OnConnectLose() {

                                }
                            });
                    fingerOximeter.Start();
                    fingerOximeter.SetWaveAction(true);
                }
            }
        });

        handler = new CurveHandler(this);
        mDtBloodOx.setmHandler(handler);

        mBtnOpenBle.setOnClickListener(v -> {
            switch (mBtnOpenBle.getText().toString()) {
                case "开始扫描":
                    JamBoHelper.getInstance().scanOxiMeterDevice();
                    break;
                case "停止扫描":
                    JamBoHelper.getInstance().stopScan();
                    break;
            }
        });

    }

    /**
     * 暂停记录，暂停绘制波纹图
     */
    private void pauseRecord() {
        if (mDtBloodOx != null) {
            mDtBloodOx.pauseDraw();
        }
        if (fingerOximeter != null) {
            fingerOximeter.Stop();
        }
    }

    private static class CurveHandler extends Handler{
        private WeakReference<OxiMeterActivity> weakReference;

        public CurveHandler(OxiMeterActivity activity) {
            this.weakReference = new WeakReference<>(activity);
        }

        @Override
        public void handleMessage(@NonNull Message msg) {
            OxiMeterActivity activity = weakReference.get();
            if (activity != null) {
                switch (msg.what) {
                    case MSG_DATA_SPO2_WAVE:
                        //波形图数据，开始绘制
                        activity.mDtBloodOx.startDraw();
                        break;
                    case MSG_DATA_SPO2_PARA:
                        //波形参数
                        if (msg.getData() != null) {
                            SpO2Data spO2Data = (SpO2Data) msg.getData().getSerializable("data");
                            if (spO2Data.getSpO2() > 0 && spO2Data.getPR() > 0) {
                                if (!spO2Data.isStatus()) {
                                    sendEmptyMessage(MSG_PROBE_OFF);
                                } else {
                                    activity.mTvValueRealtime.setText(spO2Data.toString());
                                }
                            }
                        }
                        break;
                    case MSG_PROBE_OFF:
                        activity.mDtBloodOx.cleanWaveData();
                        activity.pauseRecord();
                        if (activity.oximeterReader != null) {
                            activity.oximeterReader.close();
                        }
                        if (activity.oximeterWriter != null) {
                            activity.oximeterWriter.close();
                        }
                        JamBoHelper.getInstance().onBTStateChanged(BleState.SCAN_START);
                        break;
                }
            }
        }
    }

    @Override
    protected void onDestroy() {
        JamBoHelper.getInstance().destroy();
        super.onDestroy();
    }
}

File: app/src/main/java/com/jianbao/jamboblesdk/OxiMeterActivity.kt
package com.jianbao.jamboblesdk

import android.bluetooth.BluetoothDevice
import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.creative.FingerOximeter.FingerOximeter
import com.jianbao.jamboble.BaseBleHelper.Companion.MESSAGE_TIMEOUT
import com.jianbao.jamboble.BleHelper
import com.jianbao.jamboble.BleState
import com.jianbao.jamboble.callbacks.BleDataCallback
import com.jianbao.jamboble.callbacks.IBleStatusCallback
import com.jianbao.jamboble.data.BTData
import com.jianbao.jamboble.data.OximeterData
import com.jianbao.jamboble.data.SpO2Data
import com.jianbao.jamboble.device.BTDeviceSupport
import com.jianbao.jamboble.device.oximeter.FingerOximeterCallback
import com.jianbao.jamboble.device.oximeter.OximeterDevice
import com.jianbao.jamboble.device.oximeter.OximeterReader
import com.jianbao.jamboble.device.oximeter.OximeterWriter
import com.jianbao.jamboble.draw.DrawThreadNW
import java.lang.ref.WeakReference

class OxiMeterActivity : AppCompatActivity() {
    private val mDtBloodOx by lazy(LazyThreadSafetyMode.NONE) { findViewById<DrawThreadNW>(R.id.dt_blood_ox) }
    private val mTvValueRealtime by lazy(LazyThreadSafetyMode.NONE) { findViewById<TextView>(R.id.tv_value_realtime) }
    private val mBtnOpenBle by lazy(LazyThreadSafetyMode.NONE) { findViewById<Button>(R.id.btn_open_ble) }

    private var mFingerOximeter: FingerOximeter? = null
    private var mOximeterWriter: OximeterWriter? = null
    private var mOximeterReader: OximeterReader? = null

    private lateinit var mHandler: Handler

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_oxi_meter)

        mBtnOpenBle.setOnClickListener {
            BleHelper.instance.doSearch(this, BTDeviceSupport.DeviceType.OXIMETER)
        }
        //通用数据回调
        BleHelper.instance.setDataCallBack(
            object : BleDataCallback {
                override fun onBTStateChanged(state: BleState) {
                    when (state) {
                        //未开启蓝牙
                        BleState.NOT_FOUND -> {
                            mTvValueRealtime.text = "请打开蓝牙"
                        }
                        //正在扫描
                        BleState.SCAN_START -> {
                            mTvValueRealtime.text = "开始扫描..."
                        }
                        //连接成功
                        BleState.CONNECTED -> {
                            mTvValueRealtime.text = "连接设备成功"
                        }
                        //长时间未搜索到设备
                        BleState.TIMEOUT -> {
                            mTvValueRealtime.text = "超时"
                        }
                    }
                }

                override fun onBTDataReceived(data: BTData?) {
                    if (data is OximeterData) {
                        val btDevice = BleHelper.instance.getConnectedDevice()
                        if (btDevice is OximeterDevice) {
                            btDevice.oximeterHelper.also {
                                it.addBuffer(data.data)
                            }
                        }
                    }
                }

                override fun onLocalBTEnabled(enabled: Boolean) {
                    //蓝牙授权失败
                }

            })

        BleHelper.instance.setBleStatusCallback(
            object : IBleStatusCallback {
                override fun onBTDeviceFound(device: BluetoothDevice?) {
                }

                override fun onNotification() {
                    BleHelper.instance.getConnectedDevice().also {
                        if (it is OximeterDevice) {
                            it.oximeterHelper?.also { helper ->
                                mOximeterReader = OximeterReader(helper)
                                mOximeterWriter = OximeterWriter(helper)
                                mFingerOximeter = FingerOximeter(
                                    mOximeterReader,
                                    mOximeterWriter,
                                    FingerOximeterCallback(mHandler, mDtBloodOx)
                                )
                                mFingerOximeter?.Start()
                                mFingerOximeter?.SetWaveAction(true)

                            }

                        }
                    }
                }
            }
        )

        mHandler = CurveHandler(this)
        mDtBloodOx.setmHandler(mHandler)
    }

    /**
     * 暂停记录，暂停绘制波纹图
     */
    private fun pauseRecord() {
        if (mDtBloodOx != null) {
            mDtBloodOx.pauseDraw()
        }
        if (mFingerOximeter != null) {
            mFingerOximeter!!.Stop()
        }
    }

    override fun onDestroy() {
        //释放资源
        BleHelper.instance.destroy()
        super.onDestroy()
    }

    class CurveHandler(activity: OxiMeterActivity) : Handler() {
        private val mWeakReference = WeakReference(activity)

        override fun handleMessage(msg: Message) {
            mWeakReference.get()?.also { act ->
                when (msg.what) {
                    FingerOximeterCallback.MSG_DATA_SPO2_WAVE -> {
                        //波形图数据，开始绘制
                        act.mDtBloodOx.startDraw()
                    }
                    FingerOximeterCallback.MSG_DATA_SPO2_PARA -> {
                        //波形参数
                        if (msg.data != null) {
                            val data: SpO2Data? = msg.data.getSerializable("data") as SpO2Data?
                            data?.also {
                                if (it.spO2 > 0 && it.pr > 0) {
                                    //nStatus探头状态 ->true为正常 false为脱落
                                    if (!it.isStatus) {
                                        sendEmptyMessage(FingerOximeterCallback.MSG_PROBE_OFF)
                                    } else {
                                        //显示正在测量的实时数据
                                        act.mTvValueRealtime.text = it.toString()
                                    }

                                }
                            }

                        }
                    }

                    FingerOximeterCallback.MSG_PROBE_OFF -> {
                        act.mDtBloodOx.cleanWaveData()
                        act.pauseRecord()
                        if (act.mOximeterReader != null) {
                            act.mOximeterReader?.close()
                        }
                        if (act.mOximeterWriter != null) {
                            act.mOximeterWriter?.close()
                        }

                        //需要等待自动关闭
                        BleHelper.instance.onBTStateChanged(BleState.SCAN_START)
                    }
                    MESSAGE_TIMEOUT -> {
                        BleHelper.instance.onBTStateChanged(BleState.TIMEOUT)
                    }
                    else -> {

                    }
                }

            }
        }
    }
}